pkg/services/ngalert/notifier/channels/pagerduty.go

```diff
@@ -0,0 +1,219 @@
+		AutoResolve:   autoResolve,
+		Class:         class,
+		Component:     component,
+		Group:         group,
+		CustomDetails: details,
+		Summary:       summary,
+		tmpl:          t,
+		externalUrl:   u,
+		log:           log.New("alerting.notifier." + model.Name),
+	}, nil
+}
+
+// Notify sends an alert notification to PagerDuty
+func (pn *PagerdutyNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	alerts := types.Alerts(as...)
+	if alerts.Status() == model.AlertResolved && !pn.AutoResolve {
+		pn.log.Info("Not sending a trigger to Pagerduty", "status", alerts.Status(), "auto resolve", pn.AutoResolve)
+		return true, nil
+	}
+
+	msg, eventType, err := pn.buildPagerdutyMessage(ctx, alerts, as)
+	if err != nil {
+		return false, errors.Wrap(err, "build pagerduty message")
+	}
+
+	body, err := json.Marshal(msg)
+	if err != nil {
+		return false, errors.Wrap(err, "marshal json")
+	}
+
+	pn.log.Info("Notifying Pagerduty", "event_type", eventType)
+	cmd := &models.SendWebhookSync{
+		Url:        pagerdutyEventAPIURL,
+		Body:       string(body),
+		HttpMethod: "POST",
+		HttpHeader: map[string]string{
+			"Content-Type": "application/json",
+		},
+	}
+	if err := bus.DispatchCtx(ctx, cmd); err != nil {
+		pn.log.Error("Failed to send notification to Pagerduty", "error", err, "body", string(body))
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (pn *PagerdutyNotifier) buildPagerdutyMessage(ctx context.Context, alerts model.Alerts, as []*types.Alert) (*pagerDutyMessage, string, error) {
+	key, err := notify.ExtractGroupKey(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+
+	eventType := pagerDutyEventTrigger
+	if alerts.Status() == model.AlertResolved {
+		eventType = pagerDutyEventResolve
+	}
+
+	data := notify.GetTemplateData(ctx, &template.Template{ExternalURL: pn.externalUrl}, as, gokit_log.NewNopLogger())
+	var tmplErr error
+	tmpl := notify.TmplText(pn.tmpl, data, &tmplErr)
+
+	details := make(map[string]string, len(pn.CustomDetails))
+	for k, v := range pn.CustomDetails {
+		detail, err := pn.tmpl.ExecuteTextString(v, data)
+		if err != nil {
+			return nil, "", errors.Wrapf(err, "%q: failed to template %q", k, v)
+		}
+		details[k] = detail
+	}
+
+	msg := &pagerDutyMessage{
+		Client:      "Grafana",
+		ClientURL:   pn.externalUrl.String(),
+		RoutingKey:  pn.Key,
+		EventAction: eventType,
+		DedupKey:    key.Hash(),
+		Links: []pagerDutyLink{{
+			HRef: pn.externalUrl.String(),
+			Text: "External URL",
+		}},
+		Description: getTitleFromTemplateData(data), // TODO: this can be configurable template.
+		Payload: &pagerDutyPayload{
+			Component:     tmpl(pn.Component),
+			Summary:       tmpl(pn.Summary),
+			Severity:      tmpl(pn.Severity),
+			CustomDetails: details,
+			Class:         tmpl(pn.Class),
+			Group:         tmpl(pn.Group),
+		},
+	}
+
+	if hostname, err := os.Hostname(); err == nil {
+		// TODO: should this be configured like in Prometheus AM?
+		msg.Payload.Source = hostname
```

Comment:
Not sure we should include this. How would this look like on our cloud?

Reply:
Cloud/Prometheus alertmanager allows you to configure a source. In Grafana we used the hostname as source. So should we also make it configurable here?